java/src/main/native/src/CuFileJni.cpp

```diff
@@ -0,0 +1,321 @@
+  std::string error = cuFileGetErrorString(status.err);
+  if (IS_CUDA_ERR(status)) {
+    error.append(".").append(GetCuErrorString(status.cu_err));
+  }
+  return error;
+}
+
+/**
+ * @brief RAII wrapper for the cuFile driver.
+ */
+class cufile_driver {
+public:
+  /** @brief Construct a new driver instance by opening the cuFile driver. */
+  cufile_driver() {
+    auto const status = cuFileDriverOpen();
+    if (status.err != CU_FILE_SUCCESS) {
+      CUDF_FAIL("Failed to initialize cuFile driver: " + cuFileGetErrorString(status));
+    }
+  }
+
+  // Disable copy (and move) semantics.
+  cufile_driver(cufile_driver const &) = delete;
+  cufile_driver &operator=(cufile_driver const &) = delete;
+
+  /** @brief Destroy the driver instance by closing the cuFile driver. */
+  ~cufile_driver() { cuFileDriverClose(); }
+};
+
+/** @brief RAII wrapper for a device buffer used by cuFile. */
+class cufile_buffer {
+public:
+  /**
+   * @brief Construct a new cuFile buffer.
+   *
+   * @param device_pointer Pointer to the device buffer.
+   * @param size The size of the allocated device buffer.
+   * @param register_buffer Whether to register the buffer with cuFile. This should only be set to
+   * true if this buffer is being reused to fill a larger buffer.
+   */
+  cufile_buffer(void *device_pointer, std::size_t size, bool register_buffer = false)
+      : device_pointer_{device_pointer}, size_{size}, register_buffer_{register_buffer} {
+    if (register_buffer_) {
+      auto const status = cuFileBufRegister(device_pointer_, size_, 0);
+      if (status.err != CU_FILE_SUCCESS) {
+        CUDF_FAIL("Failed to register cuFile buffer: " + cuFileGetErrorString(status));
+      }
+    }
+  }
+
+  // Disable copy (and move) semantics.
+  cufile_buffer(cufile_buffer const &) = delete;
+  cufile_buffer &operator=(cufile_buffer const &) = delete;
+
+  /** @brief Destroy the buffer by de-registering it if necessary. */
+  ~cufile_buffer() {
+    if (register_buffer_) {
+      cuFileBufDeregister(device_pointer_);
+    }
+  }
+
+  /**
+   * @brief Get the pointer to the underlying device buffer.
+   *
+   * @return Pointer to the device buffer.
+   */
+  void *device_pointer() const { return device_pointer_; }
+
+  /**
+   * @brief Get the size of the underlying device buffer.
+   *
+   * @return The size of the device buffer.
+   */
+  std::size_t size() const { return size_; }
+
+private:
+  /// Pointer to the device buffer.
+  void *device_pointer_;
+  /// Size of the device buffer.
+  std::size_t size_;
+  /// Whether to register the buffer with cuFile.
+  bool register_buffer_;
+};
+
+/** @brief RAII wrapper for a file descriptor and the corresponding cuFile handle. */
+class cufile_file {
+public:
+  /**
+   * @brief Construct a file wrapper.
+   *
+   * @param path Absolute path to the underlying file.
+   */
+  cufile_file(char const *path) {
+    file_descriptor_ = open(path, O_CREAT | O_RDWR | O_DIRECT, 0644);
```

Comment:
It feels a little odd that we will create a file, even in the case of a read.  I am also concerned about the security aspects of `0644` being the only option for file permissions. It is not the end of the world if the parent directory is properly locked down, but I would much rather see the permissions exposed in some form to let the end user pick. I would also like to see the flags be set by the called of `cufile_file` so they are specific to the operation being performed.

Reply:
The main thing is we need to set `O_DIRECT`. Probably it's better to have two separate factory methods, one for reading one for writing?

Reply:
Created two separate factory methods for reading and writing.